### PR TITLE
P4 3722 auto repair report ingress feature

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
This feature change adds support for the service to attempt to auto repair the reports ingress should it fail due to missing reports files

In short it will now go back to the day after the last successful import and use that as a starting point and continue to the day before date of execution.

It has a env var toggle so we can switch it off should we need e.g. genuine reason for it to missing data on a give date.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/tasks/ImportReportsDataTask.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.pecs.jpc.config.TimeSource
+import uk.gov.justice.digital.hmpps.pecs.jpc.service.MonitoringService
+import uk.gov.justice.digital.hmpps.pecs.jpc.service.reports.ImportReportsService
+import java.time.LocalDate
+
+@Component
+class ImportReportsDataTask(
+  @Value("\${IMPORT_REPORTS_BACKDATE_ENABLED}") private val backDateEnabled: Boolean = false,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/tasks/ImportReportsDataTask.kt:12]: This task replaces the two separate tasks for previous days moves, events, journeys task and previous days profiles and peoples task.

File: helm_deploy/calculate-journey-variable-payments/templates/_envs.tpl [hunk 2]
```diff
@@ -145,4 +145,7 @@ env:
   - name: FEATURE_FLAG_INCLUDE_RECONCILIATION_MOVES
     value: "{{ .Values.env.FEATURE_FLAG_INCLUDE_RECONCILIATION_MOVES }}"
 
+  - name: IMPORT_REPORTS_BACKDATE_ENABLED
```
Comment (matt-crowson-moj) [inline helm_deploy/calculate-journey-variable-payments/templates/_envs.tpl:148]: So it occurred to me we need to switch this off as there may well situations where we miss days e.g. a major service outage beyond a day whereby we resort back to a paper based system ergo potentially no electronic move data.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/tasks/TasksSchedulingConfiguration.kt [hunk 3]
```diff
@@ -26,8 +24,7 @@ class TasksSchedulingConfiguration(
   @Scheduled(cron = "\${CRON_IMPORT_REPORTS}")
   @SchedulerLock(name = "importReports")
   fun previousDaysReportsTask() {
-    previousDaysMoves.execute()
-    previousDaysPeopleAndProfiles.execute()
+    importReportsDataTask.execute()
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/tasks/TasksSchedulingConfiguration.kt:27]: We now only have one task responsible for importing all data.  This is to make sure all files are present prior to importing.  When split into two the two tasks do not know about each other.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/tasks/ImportReportsDataTask.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+          return
+        }
+      }
+
+      service.importAllReportsOn(this)
+    }
+  }
+
+  private fun backdateReportsFrom(from: LocalDate) {
+    from.datesUntil(timeSource.date()).forEach { service.importAllReportsOn(it) }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/tasks/ImportReportsDataTask.kt:35]: Basically repeats until the present day (the present day is exclusive here, it does not run for this date).

**Review by Thource — APPROVED** [review]

**Pull request merged**